### PR TITLE
[WIP] Adds programmable nanites to robotics

### PR DESCRIPTION
**Pull request by XDTM**
:cl: XDTM
add: Added programmable nanites to robotics!
add: Idle nanites are the basis of the nanite system: pattern (programmed) nanites will draw from idle nanites to activate their effects. Idle nanites are cheap and do not decay over time like normal reagents.
add: Pattern nanites are produced in single-use injectors; from there, they can be programmed to customize their functionality.
add: Nanites can be given codes to activate, deactivate, self destruct or trigger. These can be used through a nanite control remote.
add: Nanites can be given an activation delay, and a timer with vcarying effects, ranging from triggering periodically to self-deactivating after a given time.
add: Nanite designs can be researched through techwebs.
/:cl:

This PR is still missing some polishing, machinery, RnD integration, mapping, and maybe adding some more nanite types (accepting suggestions!).
I'm posting it early to get feedback (and also because i feel pretty hyped about this).

This system is very dependent on trust: nanites are powerful things, and can be a cheap endless mix of healing chems. However, accepting nanites in your body means that the roboticist could sneak in some shock nanites in your bloodstream during an injection, meaning that he could stun you with a button press any time he feels like it. 

Plus, command has good reasons to enforce nanite inoculation:
With monitoring nanites, the CMO/security can finally force sensors on people! And they won't be able to turn them off this time around.
Population control nanites like pacifying, sleeping, shocking, or paralyzing can help quell a riot. What, you don't want nanites? _Do you have something to hide?_
Regeneration/antistun nanites are very handy for anyone.

Any feedback is appreciated!

### CODE REVIEW

**Comment by Armhulen** [comment]
Obligatory Nanomachines, Son!

**Comment by PKPenguin321** [comment]
Nice

**Comment by Kromgar** [comment]
Carcerand nanites that can be injected into someones blood and if they refuse orders injects chemicals into their blood that kills them. 


Yes?

**Comment by Iamgoofball** [comment]
>chem implant re-add attempt #3454

change the channel

**Comment by davethwave** [comment]
Seems interesting.

**Comment by Shadowflame909** [comment]
Goof, I think you linked the wrong thing. That isn't anything. Also, what chemical implants got removed? But it was a good way to keep people in round! Chemical implant them with a potassium + water explosion. Tell the traitor to act good or Boom. It created some interesting scenarios boss.

File: code/modules/research/nanites/nanite_remote.dm
Comment (JJRcop) [inline code/modules/research/nanites/nanite_remote.dm]: Don't forget this

**Comment by MimicFaux** [comment]
Can they be dispersed with smoke?

**Comment by YakumoChen** [comment]
@Shadowflame909 @Iamgoofball they didn't remove chem implants, just self-triggering ones. Which, like this, can't be self-triggered.

**Comment by davethwave** [comment]
Self triggered chem implants=pill in tooth surgery

**Comment by YakumoChen** [comment]
which unfortunately is a surgery and not a spammable DIY self-triggered implant

**Comment by davethwave** [comment]
true enough

**Comment by Iamgoofball** [comment]
>and can be a cheap endless mix of healing chem

yeah no

**Comment by YakumoChen** [comment]
surgery is also a cheap way of getting massive buffs but someone else has to install them in you, like this.

**Comment by XDTM** [comment]
> Can they be dispersed with smoke?

No, but electric shocks will kill a large amount of nanites, and EMPs can scramble the programming of pattern nanites, toggling them, destroying them, or even radically changing their effect! So with a little luck you can turn those kill nanites into nice heal nanites. Or viceversa.

>>and can be a cheap endless mix of healing chem

>yeah no

The price is that you could be getting a killswitch along with healing, nanites are risky! And doing it yourself is not 100% safe either, because you could get EMPd or someone could start a chain of infective nanites, which would give you the dangerous nanites anyway, that would be fueled by your own supply.

**Comment by ShadowDeath6** [comment]
I think the question was if they can be given to people via smoke

also I'm not a big fan as it's very reminiscent of the pre-nerf heal virus stacking you could do, depending on how strong the passive stuff is you would need to make them VERY vulnerable to outside sources, since EMPs are relatively rare ime and the robo releasing a bad nanite strain I would assume is relatively easy to track down a la virology

tbh I'd rather science not infringe on medical's territory, especially with the passive anti-stun / regeneration thing, surgery exists for a reason and why even waste time with surgery when you can just get nanites instead

**Comment by Ratvar** [comment]
So, meth/potass-water/synthflesh which you can't activate on your own unless you are roboticist and also antags have  more reasons to use EMP? What's the, uh, bonus, exactly.
Nanite bees.
Make nanites repair synths/robolimbs?
Nanite type that eats your nutriniments to regenerate neutral pile? +gonna eat ur ass if you don't eat much

**Comment by XDTM** [comment]
Yeah i'll definitely convert existing nanites to this system eventually

**Comment by Tlaltecuhtli** [comment]
ntnet integration?

**Comment by kevinz000** [comment]
@KorPhaeron are you ready for me to turn this into virology 2.0?

**Comment by kevinz000** [comment]
@Tlaltecuhtli :joy:

**Comment by AlexTheSysop42** [comment]
> Yeah i'll definitely convert existing nanites to this system eventually

Speaking of which, add borging nanites to the system (still only be obtainable through xenobio, maybe in a way that relies less on RNG, like adding it to black slimes)

**Comment by Nabski89** [comment]
Hey it's happening at last!

"cheap endless mix of healing chems" hasn't this already been axed like three times in virology?

**Comment by XDTM** [comment]
The difference here is that once you know a virus heals you, you know you're safe, while nanites can strike anytime. Even if the initial injection was clean, someone could still jab you with some bad nanites afterwards.

For combat purposes, EMPs act as a counter mechanic to regen nanites, as they can completely reverse function and start killing you. Shocks are a less available but still effective alternative.

Last but not least the automatic nanite generation from replicating nanites is limited; if you want to run several kinds of healing at once, either you regularly resupply your idle nanites, or you only turn them on when necessary, which leaves a window to get stunned before you can activate the antistun nanites.

**Comment by Jazaen** [comment]
Are you going to add said infectious spreading nanites to traitor uplink?

**Comment by AlexTheSysop42** [comment]
Looks like infective nanites are producible normally

**Comment by XDTM** [comment]
They're probably going to end up fairly late in the techwebs tree. Traitor roboticist will either have a whole kit that'll include some infective nanites as well, or will be able to unlock the recipes early through emagging or a design disk.

**Comment by Kromgar** [comment]
If goof hates it you know it's a good pr

File: code/modules/research/nanites/nanite_printer.dm
Comment (AlexTheSysop42) [inline code/modules/research/nanites/nanite_printer.dm]: `"Augmentation Nanites",`

**Review by xenon-difluoride — CHANGES_REQUESTED** [review]

**Comment by pubby** [comment]
Adds several invisible passive buffs: 

- Passive adrenals
- Passive armor
- Passive blood regen
- Passive temperature regulation
- Passive healing

Adds several deathmixes of the "get pricked by a pen and die" nature.

Has large overlaps with virology effects, even so far as including this:

```
/datum/design/nanites/infective
 	name = "Infective Nanites"
 	desc = "Coordinates nanite movement, making nanites able to infect nearby potential hosts."
```

>However, accepting nanites in your body means that the roboticist could sneak in some shock nanites in your bloodstream during an injection

I suppose, but can't you just use the nanite scanner on yourself to know what nanites are in you?

>Population control nanites like pacifying, sleeping, shocking, or paralyzing can help quell a riot. What, you don't want nanites? Do you have something to hide?

But people can shock themselves pretty easily to remove the nanites?

---

I'm probably misunderstanding lots of this, but the first impression is :-1: from me.

**Comment by Dorsisdwarf** [comment]
You say the downside of nanites is "but the roboticist might sneak bad ones in", but nanites can be self-injected unlike surgery, so what if you're the roboticist/guy who has the nanites? What balances them then?

**Comment by XDTM** [comment]
Some counterpoints:
- Not everyone has a nanite scanner to check their own nanites. Normal chem scans don't detect their type. My first idea was that there'd only be the roundstart ones in robotics, but i need to find a way to strike a good balance between "irreplaceable" and "free to print".
- The instadeath nanites are experimental and are either going to be very lategame, traitor-only, nuke op only, or admin-only. I'm just coding everything that seems like a nice idea, then i'm going to add only the ones that can work in a balanced way.
- Nanites have a price! You could stock up on a nanite supply, but if you have several things going on it's going to run dry fast. Passive generation is limited, as you can't just stack multiple replicating nanites.
- To add to the last one, you already can stock up on several beneficial drugs roundstart if you feel like it, especially with recipe macros being a thing now. Nanites will be gated by techweb progress, which will make them more prominent only later in the round.
- Self-shocking is probably not too attractive an option, but if it's too easy i could add a nanite type that throws an alarm/sends a nanite code if it happens. Roboticists could optionally include it, but it's extra effort for rare payoff.
- Being the roboticist with the nanites is akin to being the virologist with the heal virus. Yes, you can trust yourself, but the system isn't (or shouldn't be) op enough to become a huge issue, since as i mentioned if you run multiple nanites you're going to run out of idles very quickly.
- The stuff about someone injecting/spreading malignant nanites is still valid if someone sneaks in and prints them. Either you stay in the lab, doing nothing with your cool nanites and keeping an eye on the printer, or you roam around looking for valids or whatever while traitor mcgee prepares sleepypen 2.0 just for you.

**Comment by XDTM** [comment]
I'm going to change the scanner form a handheld device into a static machine you need to get into. (And a buddy has to read the scan).

Another half-idea is to make this machine also the main way of inoculating nanites; instead of having replicating nanites as a programmable type, idle nanites would replicate by themselves, and only be obtainable with this machine. You'd still need someone to push the nanite button while you're inside, again with the possibility that they could inject programmed nanites along with the idles.
Also it would be a clearer demarcation between nanited and non-nanited, since there'd be no longer the case where someone sips a glass of nanites, injects some regeneration nanites, runs out of idles and is back to square one.

And with _that_ i'm half-thinking about decoupling nanites from reagents altogether, maybe into a component system, since they barely react with other reagents anyway and i keep having to override reagent stuff to make this work. It probably might make things easier down the line.

**Comment by lordsaladin** [comment]
Consume. Enhance. Replicate. The syndie nanites should be based on SIVA, sentient nanites should be a random event antag.

**Review by Garen7 — CHANGES_REQUESTED** [review]
Hey some of the sprites in icons/obj/device.dmi have black stuff on them now and I don't think you meant to do that.

**Comment by Grazyn** [comment]
Instead of adding reskinned, easier to use implants to robotics, why not give them their original implants back? All those implants that were added during the years are now taking dust in the Medbay lathe because doctors never print them

**Comment by Ratvar** [comment]
Let dem malf AI disperse and control speshul snowflaek nanites

**Comment by Tlaltecuhtli** [comment]
@Grazyn robitics do have the implants they dont have the shitty cyber organs
 if it exist doesnt mean people would actually use it

**Comment by XDTM** [comment]
Consider all this code obsolete i'm porting it to components

**Comment by Nabski89** [comment]
How do you feel about having some kind of visual feedback of who does or doesn't have nanites. I'm thinking something like a neck or glove item so that players can see if you're wearing them. For a lore reason it's the transmitter that keeps them active. This also gives the person with them some way to check the status of them.
image

Personally I like the hand visual more, but gloves are a much more important slot than neck.

**Comment by XDTM** [comment]
I was planning on only putting a sec/diaghud indicator for them.  However, what i could do is add an item that boosts nanite regeneration (maybe gloves like you said), but is very visible.

**Comment by Spaceinaba** [comment]
can we please have nanites that unironically harden in response to physical trauma turning your skin black so we can metal gear memes

**Comment by Spaceinaba** [comment]
also after glancing at the code I cant wait for the CMO to hold robotics hostage until they release infectious suit sensor nanites so they can find everyone because it sounds hilarious

**Comment by AlexTheSysop42** [comment]
@Grazyn exofabs can print implants

**Comment by Fox-McCloud** [comment]
These seem really awesome, but I fear their potential for powergaming and abuse will be on the verge of gamebreaking.

The idea is really stellar though 👍 if it can be properly balanced.

**Comment by Bawhoppen** [comment]
I have a strong inkling that this will be one of those things that is OP when it gets merged, becomes the new meta, slowly gets even stronger through continual powercreep PRs, then eventually has a nerf/removal PR made, which has tons of people up in arms trying to claim it's not actually OP, which eventually ends in a merge by oranges, at the behest of many butthurt science mains who demand him to be removed as maintainer. 

Sounds fun, let's do this.

**Comment by Alek2ander** [comment]
Nanite memes not implemented: grey goo.
There should be a risk of spawning swarmers and/or blob, otherwise this is hugbox nanotechnology.

**Comment by Grazyn** [comment]
@Bawhoppen you don't have to imagine it, this is a straight up powercreeped version of implants. Everything about nanites could've been reworked into implants, but implants have a big downside which is that you need someone else to make surgery, and nanites bypass it completely. "Can be countered by EMP" isn't really a downside, especially when there isn't a clear tell that they have nanites (unless you have huds), EMP stuff is rare and then again you're not even guaranteed to neutralise them with a single EMP

**Comment by Spaceinaba** [comment]
idk if u guys can read but being shocked also kills the nanites inside people
idk what this does for balance but you guys seem to think that emp was the only thing mentioned

**Comment by lordsaladin** [comment]
Look if you want to stop nanites  cryosyringe someone with any shock causing chems.

**Comment by davethwave** [comment]
Or break one of your heads windows and run into the grill. If those are not available use engineering or SEC's windows.

**Comment by Grazyn** [comment]
@Spaceinaba yeah silly me tactical shocking is so common, it's just so easy to make teslium on the fly or a tesla gun or even something simpler like tricking them into running against a shocked grille. You could even non-human them so that silicons can help by shocking doors! So many counters!

**Comment by Jazaen** [comment]
Could we get one of those airport gates thing, but it removes _all_ nanites from those passing through?

**Comment by XDTM** [comment]
Maybe not remove or it would remove all threat from danger nanites, but it could beep loudly

**Comment by SpinnerMaster** [comment]
NANOMACHINES SON!

**Pull request closed without merging**
